### ISSUE TITLE
Postgres backup not working anymore after updating to V4

### ISSUE THREAD

**Issue by BlackScreen** [tiredofit/docker-db-backup#271]
### Summary

I have updated the docker image to V4.0.4 and tried it with the Postgres DB (I have also an InfluxDB and a MariaDB but did not test them yet) - it is not working. Changed everything according to the docs.

### Steps to reproduce

My old, working docker-compose section from V3:

```
  healthchecks-backup:
    image: tiredofit/db-backup
    container_name: healthchecks-backup
    restart: always
    security_opt:
      - no-new-privileges:true
    environment:
      - DB_TYPE=pgsql
      - DB_HOST=healthchecks-db
      - DB_NAME=healthchecks
      - DB_USER=postgres
      - DB_PASS=password
      - DB_DUMP_FREQ=1440
      - DB_DUMP_BEGIN=0300
      - DB_CLEANUP_TIME=8640
      - COMPRESSION=GZ
      - CONTAINER_ENABLE_MONITORING=FALSE
      - TZ=Europe/Berlin
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - ./backups:/backup
    depends_on:
      - healthchecks-db
    networks:
      - default
```

The edited, not working docker-compose section for V4:

```
  healthchecks-backup:
    image: tiredofit/db-backup
    container_name: healthchecks-backup
    restart: always
    security_opt:
      - no-new-privileges:true
    environment:
      - DB01_TYPE=pgsql
      - DB01_HOST=healthchecks-db
      - DB01_NAME=healthchecks
      - DB01_USER=postgres
      - DB01_PASS=password
      - DEFAULT_BACKUP_BEGIN=0300
      - DEFAULT_BACKUP_INTERVAL=1440
      - DEFAULT_CLEANUP_TIME=8640
      - DEFAULT_COMPRESSION=GZ
      - CONTAINER_ENABLE_MONITORING=FALSE
      - TZ=Europe/Berlin
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - ./backups:/backup
    depends_on:
      - healthchecks-db
    networks:
      - default
```

### What is the expected *correct* behavior?

A working backups as before. I did not change anything in the DB-variables (except the "01") and did not change anything in the rest of the compose file. Do I have to use the "01" when I am just exporting one DB?

### Relevant logs and/or screenshots

```
[tiredofit/db-backup:4.0.4 01:17:54 /] $ backup-now
2023-11-10.01:17:59 [INFO] ** [db-backup01] Starting Manual Backup for db-backup01
2023-11-10.01:17:59 [NOTICE] ** [01-healthchecks-db__healthchecks] Dumping PostgresSQL database: 'healthchecks' and compressing with 'gzip'
pg_dump: error: connection to server at "healthchecks-db" (172.22.0.2), port 5432 failed: FATAL:  password authentication failed for user "postgres"
2023-11-10.01:18:00 [INFO] ** [01-healthchecks-db__healthchecks] DB Backup of 'pgsql_healthchecks_healthchecks-db_20231110-011759.sql.gz' completed successfully
2023-11-10.01:18:00 [NOTICE] ** [01-healthchecks-db__healthchecks] Generating MD5 sum for 'pgsql_healthchecks_healthchecks-db_20231110-011759.sql.gz'
2023-11-10.01:18:00 [NOTICE] ** [01-healthchecks-db__healthchecks] Backup of 'pgsql_healthchecks_healthchecks-db_20231110-011759.sql.gz' created with the size of 20 bytes
2023-11-10.01:18:00 [NOTICE] ** [01-healthchecks-db__healthchecks] DB Backup for 'healthchecks' time taken: Hours: 0 Minutes: 00 Seconds: 01
2023-11-10.01:18:00 [INFO] ** [01-healthchecks-db__healthchecks] Backup 01 routines finish time: 2023-11-10 01:18:00 CET with exit code 0
2023-11-10.01:18:00 [NOTICE] ** [01-healthchecks-db__healthchecks] Backup 01 routines time taken: Hours: 0 Minutes: 00 Seconds: 01
```

### Environment

- Image version / tag:  4.0.4
- Host OS: Ubuntu 22.04 LTS

I have no idea why it is complaining about the password. I did not touch that at all. The application using that DB is working as expected. Weird. Thanks for your help! :)

**Comment by tiredofit** [comment]
Interesting. I am backing up Postgres as well and nto seeing this.
I don't have the extra `security_opts` in my compose file hwoever.

Your old compose file should work without upgrading, other than it barking for you to change.
Your new file is also in the correct format as well. I wonder what we are missing here.

You could try to crawl through it with DEBUG_MODE=TRUE which will output quite alot of output but also give you an understanding whats going with the password (search the output for `PGPASSWORD`). 

Let me know and I'll try to find a fix with what you find..

**Comment by BlackScreen** [comment]
Ok, this is officially a little weird.
I noticed - during testing - that the permissions of the backup folder and the backup files were changed.
Maybe it has something to do with this (probably not, but I thought I should mention it).

Before a backup with the new version, the permissions were:
Folder was set to `root:root` and `drwxr-xr-x`.
Files were set to `root:root` and `-rw-r--r--`.

After running the backup once it changes:
Folder is now set to `10000:10000` and `drwx------`.
Files are now set to `10000:10000` and `-rwx------`.

Is this the correct behaviour?

Concerning the issues with the Postgres backup:
I just about tried everything now. I deleted the `01`, I deleted the `security_opts`, I removed the `:ro` of `localtime`, I event reverted to all the old settings - nothing worked at all. This is also with Postgres 15.
Debug-mode did not give me any clues. There were only few occasions (3) of the search for `PGPASSWORD` and all were variable assignments. No error messages or anything like that.

I did try my other containers also. And, weirdly, the other Postgres backups were working - one Postgres 15 database and one Postgres 14 database.
Also the Redis 7 backup did work flawlessly. What didn't work were the backups of the InfluxDB (no file written at all) and the MariaDB (could not even find the DB).
All my tests were done with the old settings. I did not touch the compose file.

In further testing I tried the downgrade to 3.12.0 for the containers where the backups did not work. And in all cases everything worked flawlessly again. Without changing anything else in the compose file - only the image of the backup-container.

So, in conclusion: I think I will stay with 3.12.0 for the moment for the "problem" containers. But it would be great if the issues could be solved.
If I can help you with debugging, please let me know.

**Comment by tiredofit** [comment]
Thanks for getting back - yes, file permissions have changed and no longer write as root. This is changeable for at least the file permissions however. 

So is this all happening on one single host? I'm backing up ~600 databases with the 4.x container(s) and not exerperienced this. I wrote this in another issue and maybe this will help diagnose just whats happening here. My thoughts is that the file permissions are the root cause, but there is no reason for it.

```
I think we're going to have to go to the architecture level:

    Whats the architecture, x86_64, armv7, armv8/aarch64?
    What Operating System, if Linux, what distribution (cat /etc/os-release | grep PRETTY_NAME)
    If Linux, what Kernel version (uname -a)
    If Linux, what filesystem?
    Docker or Podman?
    What version of the above. For Docker docker info | grep "Server Version)
    If Docker what Storage driver (`docker info | grep "Storage")

Then lets try some tests.

    from outside the container, can you write a file to the logs directory and see it from inside the container. example: date > ./logs/outside_inside.txt
    from inside the container, can you write to a file in the logs directory and see it from outside the container. example date > /logs/inside_outside.txt
    from a scoped user and inside the container can you write to a file in the logs directory and see it from outside the container. example: echo "$(date)" | sudo -u dbbackup tee -a > /logs/dbbackup_inside_outside.txt

I only have access to nixOS and Arch systems and they have passed all tests for all db engines so I am curious to see what could be happening here.
```

**Comment by BlackScreen** [comment]
I remembered it incorrectly that I was using Ubuntu. I changed to Debian recently. Sorry for that.

I have gathered the info you wanted:
- Architecture: x86_64
- OS: Debian GNU/Linux 12 (bookworm)
- Kernel: Linux 6.1.0-13-amd64
- Filesystem: ext4
- Docker: 24.0.7
- Docker Storage Driver: overlay2

And this is all on one host, yes.

I tried with the latest version for the "problem" Postgres and ran it again with the debug modes above.
I could not find any indication what was wrong. And I did try the tests you mentioned - I did it with the ``backups`` directory. All worked, but I had it with ``root:root`` on all occasions.

**Comment by tiredofit** [comment]
Thanks for getting this info. We're moving forward on a couple different issues here and it looks as if there is some commonality with the overlay driver.

I added an undocumented variable today `DBBACKUP_USER and _GROUP`. Could you set the following:

`DBBACKUP_USER=root`
`DBBACKUP_GROUP=root`

And see if backups magically start to work? If so, this will remove some of the protection in place but at least get things working again while we understand fix..

**Comment by BlackScreen** [comment]
I tried that - updated the image to the latest version and those two variables. The result is the same:

```
[tiredofit/db-backup:4.0.5 01:00:05 /] $ backup-now
2023-11-11.01:00:11 [INFO] ** [db-backup01] Starting Manual Backup for db-backup01
2023-11-11.01:00:11 [WARN] ** [db-backup01-scheduler] Deprecated and unsupported variable 'DB_DUMP_BEGIN' detected - Please upgrade your variables as they will be removed in version 4.3.0
2023-11-11.01:00:11 [WARN] ** [db-backup01-scheduler] Deprecated and unsupported variable 'DB_DUMP_FREQ' detected - Please upgrade your variables as they will be removed in version 4.3.0
2023-11-11.01:00:11 [WARN] ** [db-backup01-scheduler] Legacy unsupported variable 'COMPRESSION' detected - Please upgrade your variables as they will be removed in version 4.3.0
2023-11-11.01:00:12 [NOTICE] ** [01-healthchecks-db__healthchecks] Dumping PostgresSQL database: 'healthchecks' and compressing with 'gzip'
pg_dump: error: connection to server at "healthchecks-db" (172.18.0.2), port 5432 failed: FATAL:  password authentication failed for user "postgres"
2023-11-11.01:00:12 [INFO] ** [01-healthchecks-db__healthchecks] DB Backup of 'pgsql_healthchecks_healthchecks-db_20231111-010012.sql.gz' completed successfully
2023-11-11.01:00:12 [NOTICE] ** [01-healthchecks-db__healthchecks] Generating MD5 sum for 'pgsql_healthchecks_healthchecks-db_20231111-010012.sql.gz'
2023-11-11.01:00:12 [NOTICE] ** [01-healthchecks-db__healthchecks] Backup of 'pgsql_healthchecks_healthchecks-db_20231111-010012.sql.gz' created with the size of 20 bytes
2023-11-11.01:00:12 [NOTICE] ** [01-healthchecks-db__healthchecks] DB Backup for 'healthchecks' time taken: Hours: 0 Minutes: 00 Seconds: 00
2023-11-11.01:00:12 [INFO] ** [01-healthchecks-db__healthchecks] Backup 01 routines finish time: 2023-11-11 01:00:12 CET with exit code 0
2023-11-11.01:00:12 [NOTICE] ** [01-healthchecks-db__healthchecks] Backup 01 routines time taken: Hours: 0 Minutes: 00 Seconds: 00
```

The file was written - as before - but it did not have any useable content and was very small, as expected from the error.

**Comment by tiredofit** [comment]
Thanks - I'm now starting to focus on that password issue...

Without revealing too much. Is there a space, a quote, or a special character in there?

You can try a manual backup like so based on your compose environmetn variables above:
```
      - DB01_HOST=healthchecks-db
      - DB01_NAME=healthchecks
      - DB01_USER=postgres
      - DB01_PASS=password
```

```
export PGPASSWORD=${DB01_PASS}
pg_dump -h ${DB01_HOST} -p 5432 -U ${DB01_USER} ${DB01_NAME} > backup.sql
```

**Comment by BlackScreen** [comment]
I think the passwords and the special characters there could be key here - as mentioned in the other issue #274.

I have analysed my passwords and here are the special characters in the password - aside from lowercase and uppercase letters and numbers, obviously.

The "problem" container (healthchecks) with Postgres 15 has the following special characters in the password: ``!$%*@^``
The MariaDB container with the problems (same as in the mentioned issue) has this special characters in the password: ``!%*@^``
And the token of the InfluxDB has this special character inside: ``=``
But maybe the InfluxDB one is another problem, I don't know.

They all are entered in the compose file in this way:
```
    environment:
      - DB_PASS=password_goes_here
```
The ``$`` in the one password is escaped with two ``$$`` inside the compose file. It worked before and does so with other containers (the MariaDB itself for example).

And I did try the ``pg_dump`` inside the container - it worked.

**Comment by BlackScreen** [comment]
I have tried the latest version and for the "problem" container with Postgres 15 I have now successfully made a backup.
The only "issue" for me is, that I can not set the permissions on the backups folder with a environment variable. I would like to set the folder to 755 so that I can ``ls`` it with the normal user. Maybe you could add an environment variable like ``DEFAULT_FILESYSTEM_PERMISSION`` for the folder?
I will test the other containers now.

**Comment by BlackScreen** [comment]
So, I have updated and tested all containers now - everything is working. Thanks for your work!
Maybe you can add my suggestion in the upcoming releases. I think I will close this now. :)

**Comment by tiredofit** [comment]
Sure, check out the `_FILESYSTEM_PATH_PERMISSIONS` variables - that will allow you to independently be able to change the top level directory vs the files inside.
